### ISSUE TITLE
@osresearch @marmarek :

### ISSUE THREAD

**Issue by tlaurion** [QubesOS/qubes-issues#9451]
@osresearch @marmarek :
When comparing two root snapshots per
```
[user@dom0 ~]$ cat /lib/systemd/system-shutdown/root-autosnap 
#!/bin/sh

#This permits wyng-backup to backup root-autosnap and root-autosnap-back, taken at each system shutdowns like any other QubesOS LVMs.
#This also permits to restore to different states of dom0 from Heads and compare dom0 between reboots

#TODO: backup /boot content into a LVM and apply same logic, corresponding to each dom0 snapshots
#https://github.com/tasket/wyng-backup/issues/63

#We delete the backup of last shutdown snapshot (last last shutdown)
/usr/sbin/lvremove --noudevsync --force -An qubes_dom0/root-autosnap-back || true
#We take a snapshot of root-autosnap into root-autosnap-back
/usr/sbin/lvcreate --noudevsync --ignoremonitoring -An -pr -s qubes_dom0/root-autosnap -n root-autosnap-back
#We remove root-autosnap
/usr/sbin/lvremove --noudevsync --force -An qubes_dom0/root-autosnap || true
#We create root-autosnap from root
/usr/sbin/lvcreate --noudevsync --ignoremonitoring -An -pr -s qubes_dom0/root -n root-autosnap
```

And then we compare the content of the filesystems, we see that:

- /etc/libvirt/libxl
- /etc/lvm
- /etc/xdg/adjtime
- /home
- /root
- /var

Would need to be out of root fs to be able to have a [RO QubesOS dom0 with dmverity ](https://github.com/osresearch/safeboot/blob/fd0aef40912948711986e2e4132c4ba8e30e170d/sbin/safeboot#L1075)

_Originally posted by @tlaurion in https://github.com/QubesOS/qubes-issues/issues/4371#issuecomment-761863958_

**Comment by tlaurion** [comment]
@DemiMarie approves moving those out at https://github.com/QubesOS/qubes-issues/issues/4371#issuecomment-834817505

> > @marmarek @osresearch input on [#4371 (comment)](https://github.com/QubesOS/qubes-issues/issues/4371#issuecomment-761863958) ?
> 
> Personally I would be fine with moving these off of the root filesystem.

**Comment by tlaurion** [comment]
Also was https://github.com/QubesOS/qubes-issues/issues/5777#issuecomment-1294084776

**Comment by ben-grande** [comment]
@tlaurion Can you please adapt the title? I don't think you mean to move from (out of) dom0, but move to another directory so dm-verity works?

**Comment by tlaurion** [comment]
> @tlaurion Can you please adapt the title? I don't think you mean to move from (out of) dom0, but move to another directory so dm-verity works?

Title is good. To enable dm-verity,  moving volatile/dynamic content out of dom0 filesystem would allow the it to to be verified with dm-verity hashing.

Experiment comparing snapshots between  reboot would give a fresher list of states needing to be kept out to have read only dom0 between updates and verification.

**Comment by tlaurion** [comment]
Is this GSoC material? 

Most needed.

**Comment by ben-grande** [comment]
> > [@tlaurion](https://github.com/tlaurion) Can you please adapt the title? I don't think you mean to move from (out of) dom0, but move to another directory so dm-verity works?
> 
> Title is good. To enable dm-verity, moving volatile/dynamic content out of dom0 filesystem would allow the it to to be verified with dm-verity hashing.

But it is not moving from dom0 to something out of dom0 right? It is moving "in" dom0, from root to another place. This is why I asked for the title clarification. Note, I do not fully understand dm-verity, this is just my bare understanding of it.

**Comment by tlaurion** [comment]
> > > [@tlaurion](https://github.com/tlaurion) Can you please adapt the title? I don't think you mean to move from (out of) dom0, but move to another directory so dm-verity works?
> > 
> > Title is good. To enable dm-verity, moving volatile/dynamic content out of dom0 filesystem would allow the it to to be verified with dm-verity hashing.
> 
> But it is not moving from dom0 to something out of dom0 right? It is moving "in" dom0, from root to another place. This is why I asked for the title clarification. Note, I do not fully understand dm-verity, this is just my bare understanding of it.

From referred now old safeboot
> Protecting the runtime system integrity (by optionaly booting from a read-only root with dm-verity and signed root hash)

As per op

```
And then we compare the content of the filesystems, we see that:

/etc/libvirt/libxl
/etc/lvm
/etc/xdg/adjtime
/home
/root
/var

```
This was a long time ago, didn't reassess lvm changes between snapshots, but that would give complete picture. Change QubesOS configs, snapshot, assess. 

Heads recently adapted its root hashing feature tu support dom0 lvm hashing in standard installation, which can be simulated per commands in op. 
https://github.com/linuxboot/heads/pull/2067. This feature aims to hash everything in standard OS libs and binaries path to act as a HIDS Ala tripwire, and can be activated to check root hashes on boot, without expecting dm-verity to be possible (dm-verity is a fast hashing mechanism for volumes expected to not change between boot).


Files in dom0 changes without OS base changing, this is related but not limited to dom0 configs, logs, and other artifacts that should not be in dom0 in the goal of having a dom0 that is RO, prerequisite of dm-verity. 

To me title is good and references are sound. It's about moving irrelevant runtime changes out of dom0, this issue gives guidelines to investigate between lvm snapshots to see what needs to be done.

**Comment by marmarek** [comment]
> It's about moving irrelevant runtime changes out of dom0, this issue gives guidelines to investigate between lvm snapshots to see what needs to be done.

No, it isn't moving them **out of dom0**, it about possibly moving them outside of dom0 root partition, **into another dom0 partition**, but still dom0. There is definitely no desire to move system configuration out of dom0.
The above list kinda looks like it isn't even about much software change, but partition layout. `/var`, `/root`, `/home` and even `/etc` are supposed to be mutable, and persistent. Making them completely read-only is undesirable, but moving to a separate partition may be a good idea. Technically, large part of `/etc` also doesn't change much, but the changing parts are not limited to what you listed - for example when user change their password, `/etc/shadow` will be modified (and such modification must survive). Similarly, enabling/disabling repositories (both for dom0 updates and for templates) will modify some files in /etc. There are more examples like this.

**Comment by marmarek** [comment]
As for /etc, AFAIK openSUSE is attempting an interesting thing: move default configuration as much as possible into /usr (/usr/etc, usr/lib etc) - so, /etc contains only configuration specific to the current installation. I've seen somebody telling they want to go as far as the fresh installation should have almost empty /etc (except files explicitly created by the installer, like /etc/fstab). It would be interesting in this context, but I'm not sure how far they are on this way right now, last time I checked fresh openSUSE installation had still quite full /etc.

**Comment by tlaurion** [comment]
> No, it isn't moving them out of dom0, it about possibly moving them outside of dom0 root partition, into another dom0 partition, but still dom0.

Right. LogVM is already into another issue. Qubesos, Xen related changes should be off of dom0 fs. As for changes related to dom0 rootfs (/etc) I think it's totally fine to consider this a change of dom0 rootfs. If something changed my password, not me, I would definitely want a binary changed/not changed fast hashing mechanism to tell me so as early as possible, ie dm-verity 's job.

**Comment by marmarek** [comment]
> As for changes related to dom0 rootfs (/etc) I think it's totally fine to consider this a change of dom0 rootfs. If something changed my password, not me, I would definitely want a binary changed/not changed fast hashing mechanism to tell me so as early as possible, ie dm-verity 's job.

Nope. dm-verity image is fully hashed and read-only - you literally cannot mount dm-verity device read-write. _If_ going this way, it would mean that part of the system would be not user-changeable at all (unless you do your own build from source). Forcing user to do their own build of the system just to change a password is definitely not the way to go.

If you want to abuse dm-verity in Heads to check block device that is sometimes intentionally changed by the user, I guess you can try, but you wont find much support for that from my side.

**Comment by tlaurion** [comment]
I'm getting confused @marmarek. 

https://github.com/QubesOS/qubes-issues/issues/9146 is larger (path torward a measurable dom0...), where here is getting stuff out of dom0 rootfs in that goal, so that anything else (fedora silverblue or what not) can then be switched to to attain that goal. No magic will happen out of the blue, first step is listing things needed to be out of dom0 fs so that it can be measured, volatile/config specific volumes doing the rest. I do not pretend reassessing all past issues or PoC that were done before in this problem space, where things were already discussed in the goal of coming with a plan. They can be brought back if there is a desire going that path. 

There are absolutely no desire for users to build QubesOS from source to have ro fs+ rw overlay, which would keep needed changes out of dom0 in all cases, for system to run as expected. This discussion was aimed at that: what, and how to get things out of dom0 rootfs, first step being to see what changes should persist and shouldn't in that goal. And then boot ro, apply changes of overlay, and commit changes at end of reboot or other PoC conclusions that happened in the past. 

Most of the littérature and PoC were done on 3.2, where it was file backend, then safeboot recommended changes, then lvm happened and then lvm split between dom0 fs and poolfs.

Is there an interest in going forward a measured dom0 rootfs and then eventually making it ro A/B switch over? A lot changed in that spectrum from other OSes going in that direction.

**Comment by marmarek** [comment]
As I said above, moving changing parts out of root filesystem (into another partition, or some other similar solution) is a desirable move. If some mutable files are not needed at all then sure, getting rid of them is fine too, but I don't think any of the above falls into this category.
Then, the root filesystem could be mounted read-only and possibly dm-verity could be utilized to verify its integrity.
What I do not want, is applying dm-verity before mutable parts are separated (or in other words: keep some, possibly rarely changed, configuration still in the rootfs) and rebuild the merkle-tree locally after changes happen - that would mostly nullify benefits of dm-verity, as it's hard to verify if given change at the block level is legitimate or not.

I think generally we agree on the final goal, but maybe have different ideas how to get there?

**Comment by tlaurion** [comment]
https://github.com/QubesOS/qubes-issues/issues/9146#issue-2257942812

> PoC and original discussion https://groups.google.com/g/qubes-devel/c/hG93VcwWtRY

**Comment by ryrona2** [comment]
In the community, there is a desire for being able to run truly disposable qubes, ie that leave no persistent traces of their use. This would give anti-forensics capabilities some journalists and activists desire or even outright need, similar to what is offered by Tails.

Ticket: https://github.com/QubesOS/qubes-issues/issues/904

To enable this, it is very important that logs files, cache files and other volatile state that might leak activity and do not need to be persisted between reboots of dom0 is never written to disk. There is lots of such volatile state that mustn't be persisted in the "/var" folder. The only thing that should or is allowed to be persisted in this setting is configuration files and other data the user has specifically changed themselves, that do not risk containing information about activity, only about desired system configuration.

I know it is being talked about that filenames from within each qube is leaking to persistent storage in dom0 today, maybe when copying files between qubes. This is of course very problematic from an anti-forensics perspective, and a blocker today for truly disposable qubes. I can imagine that window headers and ID of plugged in devices and timestamps of various kinds also might leak to persistent storage in dom0, all also being problematic.

Just moving all folders with volatile state out so separated from read-only image is a step in the right direction, but not enough for where we want to be at the end. A better solution is to have dm-verity protected read-only image, and a writable tmpfs overlay above. And then only directories containing system configuration, and no other kind of volatile state is on a writable disk image. This means narrowing down the paths to persist way from what is suggested in this ticket, to only what is needed for QubesOS configuration and state that absolutely must survive reboots.

> The above list kinda looks like it isn't even about much software change, but partition layout. /var, /root, /home and even /etc are supposed to be mutable, and persistent. Making them completely read-only is undesirable, but moving to a separate partition may be a good idea. Technically, large part of /etc also doesn't change much, but the changing parts are not limited to what you listed - for example when user change their password, /etc/shadow will be modified (and such modification must survive). Similarly, enabling/disabling repositories (both for dom0 updates and for templates) will modify some files in /etc. There are more examples like this.

Another aspect of using dm-verity is to prevent malware persistence, even in the case dom0 has been compromised. Ie, a simple reboot should be enough to be certain no malware any longer gets executed in the context of dom0 or other highly privileged layers. Persisting all of "/etc" is likely highly problematic and defeat most of the purpose of using dm-verity, as "/etc" typically can contain scripts that gets executed during system startup, customized systemd unit files that likewise execute things during system startup, and so on and on. All ways malware can persist itself. It is important that everything that is persisted between reboots are purely system configuration, ie specification of desired system state, not executables or scripts or other ways code can be persisted. Ideally, malware should also not be able to weaken security of a Qubes OS installation beyond what is exposed through GUI configuration tools, but this means persisting a very minimal amount of configuration.

Again, separating volatile state from read-only image is a good start, but not near enough for where we need to be in the end to really benefit from dm-verity for security, ie to prevent malware persistence, as that is what dm-verity is supposed to protect against.

(However, it can be argued the user should have the right to modify their own system, and dm-verity and immutable images conflicts with this goal. It could also be argued QubesOS already provide strong protection from even being remotely infected by malware at dom0 level to begin with, and that QubesOS due to its design is unable to provide strong protection against malware persistence after dom0 compromise anyway since someone with dom0 privileges can just persist the malware in all templates instead which are mutable by design. So it can be argued maybe dm-verity should not be part of QubesOS, and that boot verification of Xen, kernel, boot arguments and initrd is enough, and the rest is protected from local attackers by disk encryption. In this chain of logic, dm-verity could be argued to be out-of-scope for QubesOS, but volatile state that mustn't survive reboot should still be written to a tmpfs overlay and never disk to implement anti-forensics capabilities.)
